[JIT] Add Reinplacing to MKLDNN Subgraphs

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53327 [WIP] torch.utils.freeze
* **#53908 [JIT] Add Reinplacing to MKLDNN Subgraphs**
* #53326 use reshape when possible in broadcasting
* #52614 Avoid dispatch overhead in call to mkldnn convolution

This adds reinplacing to MKLDNN Subgraphs so that we replace `aten::add` with `aten::add_`. Normally you would have to prove device and dtype, but we know that already, and because we have explicit broadcast nodes for other reasons we dont have to prove that the output shape of add is the same as inputs. 

Ive tested correctness on resnet, I'm going to do more extensive testing as well. When I benchmarked the "unsafe" version (always inplace) I saw average speedups of ~16% for both Single threaded and Multithreaded. I dont think the "safe" version will be far beyond; when I looked at resnet for example every `add` and `relu` were reinplaced. 

Theres some question of reusing other alias / liveness / inplacing passes in SR. I thought about it, however I didnt want to add a cross-dependency between very different parts of the code base with a bunch of different assumptions. The logic here is also covering a simpler case and does not add much complexity IMO.

Differential Revision: [D27132969](https://our.internmc.facebook.com/intern/diff/D27132969)

Commit:
Add Reinplacing to MKLDNN Subgraphs

[ghstack-poisoned]

Commit:
Update on "[JIT] Add Reinplacing to MKLDNN Subgraphs"


This adds reinplacing to MKLDNN Subgraphs so that we replace `aten::add` with `aten::add_`. Normally you would have to prove device and dtype, but we know that already, and because we have explicit broadcast nodes for other reasons we dont have to prove that the output of add is the same as inputs. 

Ive tested correctness on resnet, I'm going to do more extensive testing as well. When I benchmarked the "unsafe" version (always inplace) I saw average speedups of ~16% for both Single threaded and Multithreaded. I dont think the "safe" version will be far beyond; when I looked at resnet for example every `add` and `relu` were reinplaced. 

Theres some question of reusing other alias / liveness / inplacing passes. I thought about, however I didnt want to add a cross-dependency between two sets of people working on very different parts of the code base. The logic here is also covering a simpler case and does not add much complexity IMO. 



[ghstack-poisoned]

Commit:
Update on "[JIT] Add Reinplacing to MKLDNN Subgraphs"


This adds reinplacing to MKLDNN Subgraphs so that we replace `aten::add` with `aten::add_`. Normally you would have to prove device and dtype, but we know that already, and because we have explicit broadcast nodes for other reasons we dont have to prove that the output of add is the same as inputs. 

Ive tested correctness on resnet, I'm going to do more extensive testing as well. When I benchmarked the "unsafe" version (always inplace) I saw average speedups of ~16% for both Single threaded and Multithreaded. I dont think the "safe" version will be far beyond; when I looked at resnet for example every `add` and `relu` were reinplaced. 

Theres some question of reusing other alias / liveness / inplacing passes in SR. I thought about it, however I didnt want to add a cross-dependency between very different parts of the code base with a bunch of different assumptions. The logic here is also covering a simpler case and does not add much complexity IMO. 



[ghstack-poisoned]

Commit:
Update on "[JIT] Add Reinplacing to MKLDNN Subgraphs"


This adds reinplacing to MKLDNN Subgraphs so that we replace `aten::add` with `aten::add_`. Normally you would have to prove device and dtype, but we know that already, and because we have explicit broadcast nodes for other reasons we dont have to prove that the output shape of add is the same as inputs. 

Ive tested correctness on resnet, I'm going to do more extensive testing as well. When I benchmarked the "unsafe" version (always inplace) I saw average speedups of ~16% for both Single threaded and Multithreaded. I dont think the "safe" version will be far beyond; when I looked at resnet for example every `add` and `relu` were reinplaced. 

Theres some question of reusing other alias / liveness / inplacing passes in SR. I thought about it, however I didnt want to add a cross-dependency between very different parts of the code base with a bunch of different assumptions. The logic here is also covering a simpler case and does not add much complexity IMO. 



[ghstack-poisoned]

Krovatkin commented on torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:
```diff
@@ -39,6 +41,142 @@ c10::AliasAnalysisKind aliasAnalysisFromSchema() {
+    // Node output Value*
+    nodes_to_inplace.push_back(node);
+    auto output_liveness_end =
+        set_liveness[alias_mapping[node->outputs().at(0)]];
```
are we expecting multi-output nodes? should we assert here? and maybe a quick comment that we -re-arrange nodes to make `inputs().at(0)` reusable.

eellison replied:
there are no multi-output nodes, i can assert

Krovatkin commented on test/jit/test_freezing.py:
```diff
@@ -1762,3 +1762,103 @@ def test_conv_to_mkldnn_no_mkldnn(self):
+                # x can't be inplaced because its a return value,
+                # check that the inplacing pass doesnt try to inplace
+                # self.tensor because its always alive
+                return x * self.tensor, x
```
self.tensor would be a frozen constant?

eellison replied:
yes

Krovatkin commented on test/jit/test_freezing.py:
```diff
@@ -1762,3 +1762,103 @@ def test_conv_to_mkldnn_no_mkldnn(self):
             inp = torch.rand([4, 3, 4, 4])
             self.assertEqual(frozen(inp), mod(inp))
+
+@unittest.skipIf(not torch._C.has_mkldnn, "MKL-DNN build is disabled")
```
* do we have a test that replaces `x.relu_().relu_().relu_()` like multiple propagations? I quickly scanned through tests and didn't find one.
* do we want a test that already has an in-place op. May not be necessary since we functionalize first?

eellison replied:
sure i can add for the first case. there are already tests for the second case and correct it's not necessary since we functionalize

Krovatkin reviewed (commented):
:shipit:

eellison commented on torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:
```diff
@@ -39,6 +41,142 @@ c10::AliasAnalysisKind aliasAnalysisFromSchema() {
+    }
+  }
+
+  std::unordered_map<ValueSetPtr, Node*> set_liveness;
```
happy to separate them into different functions, but I'd have to thread around the different state that's computed in each function... I kind of thought it would be cleaner just to have one function that you read through linearly. somewhat related http://number-none.com/blow/john_carmack_on_inlined_code.html

Krovatkin replied:
> http://number-none.com/blow/john_carmack_on_inlined_code.html

hehe, hard to argue w/ Carmack on anything. It's indeed hard to track a state if it's split across a bunch of functions. In this case, I was hoping that the state is pretty clearly delineated via `alias_mapping` and `liveness` and `node_to_inplace`, so they could become members fields in a liveness pass, but I agree it's not worth it since a possible readability improvement is highly subjective.

eellison replied:
I didnt separate out the `merge_sets` function, so i went part of the way :)

eellison commented on torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:
```diff
@@ -39,6 +41,142 @@ c10::AliasAnalysisKind aliasAnalysisFromSchema() {
+  // reasons).
+
+  auto aliasDb = torch::make_unique<AliasDb>(graph);
+  std::unordered_map<Value*, ValueSetPtr> alias_mapping;
```
why would that be cleaner?

eellison replied:
i'm not sure i understand the suggestion

Krovatkin replied:
yeah, sorry, when I try to write it out, I realized that input values map to the same input set, so we can't do `std::unordered_map<Value*, ValueSet> alias_mapping;`

eellison commented on torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:
```diff
@@ -39,6 +41,142 @@ c10::AliasAnalysisKind aliasAnalysisFromSchema() {
+    for (auto it = set.second->begin(); it != set.second->end(); it++) {
+      Value* v = *it;
+      auto k = v->node()->kind();
+      if (k == prim::Constant || k == prim::ConstantMKLDNNTensor ||
```
Yes, no GetAttr

eellison replied:
I added checks that we dont have any containers which contains Tensors (getattr, lists, dicts, etc) which complicate alias analysis

eellison commented on torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:
```diff
@@ -39,6 +41,142 @@ c10::AliasAnalysisKind aliasAnalysisFromSchema() {
+  }
+
+  std::vector<Node*> nodes_to_inplace;
+  auto add_to_inplace_set = [&](Node* node) {
```
personally i like the lambdas because you dont have to thread as much state around, plus you can read the function linearly

eellison replied:
i'll try rewriting though..

eellison replied:
if i were to make this into a stand alone function i would have to do 

have 

`void add_to_inplace_set(std::unordered_map<Value*, ValueSetPtr>& alias_mapping&, std::unordered_map<ValueSetPtr, Node*>& set_liveness, std::vector<Node*>& nodes_to_inplace, Node * inplace_node)`

IMO that is not easier to follow

Commit:
Update on "[JIT] Add Reinplacing to MKLDNN Subgraphs"


This adds reinplacing to MKLDNN Subgraphs so that we replace `aten::add` with `aten::add_`. Normally you would have to prove device and dtype, but we know that already, and because we have explicit broadcast nodes for other reasons we dont have to prove that the output shape of add is the same as inputs. 

Ive tested correctness on resnet, I'm going to do more extensive testing as well. When I benchmarked the "unsafe" version (always inplace) I saw average speedups of ~16% for both Single threaded and Multithreaded. I dont think the "safe" version will be far beyond; when I looked at resnet for example every `add` and `relu` were reinplaced. 

Theres some question of reusing other alias / liveness / inplacing passes in SR. I thought about it, however I didnt want to add a cross-dependency between very different parts of the code base with a bunch of different assumptions. The logic here is also covering a simpler case and does not add much complexity IMO. 



[ghstack-poisoned]

eellison commented:
@Krovatkin updated, plz take another look

Krovatkin reviewed (approved):
:shipit: looks gut!!!

Commit:
Update on "[JIT] Add Reinplacing to MKLDNN Subgraphs"


This adds reinplacing to MKLDNN Subgraphs so that we replace `aten::add` with `aten::add_`. Normally you would have to prove device and dtype, but we know that already, and because we have explicit broadcast nodes for other reasons we dont have to prove that the output shape of add is the same as inputs. 

Ive tested correctness on resnet, I'm going to do more extensive testing as well. When I benchmarked the "unsafe" version (always inplace) I saw average speedups of ~16% for both Single threaded and Multithreaded. I dont think the "safe" version will be far beyond; when I looked at resnet for example every `add` and `relu` were reinplaced. 

Theres some question of reusing other alias / liveness / inplacing passes in SR. I thought about it, however I didnt want to add a cross-dependency between very different parts of the code base with a bunch of different assumptions. The logic here is also covering a simpler case and does not add much complexity IMO.

Differential Revision: [D27132969](https://our.internmc.facebook.com/intern/diff/D27132969)

[ghstack-poisoned]
